File: src/agents/video.py
import subprocess

from src.asset import AssetStore
from src.config import load_config


class VideoAgent:
    def __init__(self, store: AssetStore):
        self.store = store
        cfg = load_config()
        video_cfg = cfg.get("steps", {}).get("video", {})
        self.resolution = video_cfg.get("resolution", "1920x1080")
        self.fps = video_cfg.get("fps", 25)
        self.codec = video_cfg.get("codec", "libx264")

    def run(self, audio_paths: list[str]) -> str:
        self.store.log_input("video", {"audio_paths": audio_paths})

        audio_dir = self.store.audio_dir()
        video_dir = self.store.video_dir()
        audio_path = audio_dir / "full_audio.wav"
        output_path = video_dir / "video.mp4"

        width, height = self.resolution.split("x")

        cmd = [
            "ffmpeg",
            "-y",
            "-f",
            "lavfi",
            "-i",
            f"color=c=black:s={width}x{height}:r={self.fps}",
            "-i",
            str(audio_path),
            "-shortest",
            "-c:v",
            self.codec,
            "-pix_fmt",
            "yuv420p",
            "-c:a",
            "aac",
            str(output_path),
        ]

        subprocess.run(cmd, check=True)

        self.store.log_output("video", {"path": str(output_path), "status": "completed"})

        return str(output_path)

File: src/agents/video.ts

import path from "path";
import ffmpeg from "fluent-ffmpeg";
import { AssetStore } from "../asset.js";
import { loadConfig } from "../config.js";

export class VideoAgent {
    store: AssetStore;
    resolution: string;
    fps: number;
    codec: string;

    constructor(store: AssetStore) {
        this.store = store;
        const cfg = loadConfig();
        const videoCfg = cfg.steps.video;
        this.resolution = videoCfg.resolution;
        this.fps = videoCfg.fps;
        this.codec = videoCfg.codec;
    }

    async run(audioPaths: string[], thumbnailPath: string): Promise<string> {
        this.store.logInput("video", { audioPaths, thumbnailPath });
        const audioDir = this.store.audioDir();
        const videoDir = this.store.videoDir();
        const audioPath = path.join(audioDir, "full_audio.wav");
        const outputPath = path.join(videoDir, "video.mp4");
        const [width, height] = this.resolution.split("x");

        const cmd = ffmpeg()
            .input(`color=c=black:s=${width}x${height}:r=${this.fps}`)
            .inputFormat("lavfi")
            .input(audioPath);

        [thumbnailPath].filter(p => Boolean(p)).map(p => cmd.input(p));

        const filters = [
            `[0:v]copy[v]`
        ];
        [thumbnailPath].filter(p => Boolean(p)).map(() => {
            filters.length = 0;
            filters.push(`[2:v]scale=${width}:${height}[thumb]`);
            filters.push(`[0:v][thumb]overlay=0:0:enable='lte(t,5)'[v]`);
        });

        return new Promise((resolve, reject) => {
            cmd.complexFilter(filters)
                .outputOptions([
                    "-map", "[v]",
                    "-map", "1:a",
                    "-shortest",
                    "-c:v", this.codec,
                    "-pix_fmt", "yuv420p",
                    "-c:a", "aac"
                ])
                .save(outputPath)
                .on("end", () => {
                    this.store.logOutput("video", { path: outputPath, status: "completed" });
                    resolve(outputPath);
                })
                .on("error", (err) => {
                    reject(err);
                });
        });
    }
}
